src/lavinmq/mqtt/broker.cr

```diff
@@ -71,6 +77,25 @@ module LavinMQ
       end
 
       def publish(packet : MQTT::Publish)
+        # Track Sparkplug BIRTH/DEATH messages if state tracker is enabled
+        if tracker = @sparkplug_state_tracker
+          begin
+            if msg_type = Sparkplug::Validator.validate_topic(packet.topic)
+              # Parse topic to extract group_id and edge_node_id
+              if parts = Sparkplug::Validator.parse_topic(packet.topic)
```

Comment:
`#validate_topic` calls `#parse_topic` which will split the topic, so it's done twice. Maybe `#validate_topic` should take the parts instead?
```suggestion
            if parts = Sparkplug::Validator.parse_topic(packet.topic)
              if msg_type = Sparkplug::Validator.validate_topic(parts)
```
But should a validate method return a message type, not a boolean?